Gems/AtomTressFX/Assets/Shaders/HairSimulationCommon.azsli

```diff
@@ -0,0 +1,446 @@
+{
+    float r = 1.f + dot(u, v);
+    float3 n;
+
+    // if u and v are parallel
+    if (r < 1e-7)
+    {
+        r = 0.0f;
+        n = abs(u.x) > abs(u.z) ? float3(-u.y, u.x, 0.f) : float3(0.f, -u.z, u.y);
+    }
+    else
+    {
+        n = cross(u, v);  
+    }
+
+    float4 q = float4(n.x, n.y, n.z, r);
+    return NormalizeQuaternion(q);
+}
+
+// Make the inpute 4x4 matrix be identity
+float4x4 MakeIdentity()
+{
+    float4x4 m;
+    m._m00 = 1;   m._m01 = 0;   m._m02 = 0;   m._m03 = 0;
+    m._m10 = 0;   m._m11 = 1;   m._m12 = 0;   m._m13 = 0;
+    m._m20 = 0;   m._m21 = 0;   m._m22 = 1;   m._m23 = 0;
+    m._m30 = 0;   m._m31 = 0;   m._m32 = 0;   m._m33 = 1;
+
+    return m;
+}
+
+void ApplyDistanceConstraint(inout float4 pos0, inout float4 pos1, float targetDistance, float stiffness = 1.0)
+{
+    float3 delta = pos1.xyz - pos0.xyz;
+    float distance = max(length(delta), 1e-7);
+    float stretching = 1 - targetDistance / distance;
+    delta = stretching * delta;
+    float2 multiplier = ConstraintMultiplier(pos0, pos1);
+
+    pos0.xyz += multiplier[0] * delta * stiffness;
+    pos1.xyz -= multiplier[1] * delta * stiffness;
+}
+
+void CalcIndicesInVertexLevelTotal(uint local_id, uint group_id, inout uint globalStrandIndex, inout uint localStrandIndex, inout uint globalVertexIndex, inout uint localVertexIndex, inout uint numVerticesInTheStrand, inout uint indexForSharedMem, inout uint strandType)
+{
+    indexForSharedMem = local_id;
+    numVerticesInTheStrand = (THREAD_GROUP_SIZE / g_NumOfStrandsPerThreadGroup);
+
+    localStrandIndex = local_id % g_NumOfStrandsPerThreadGroup;
+    globalStrandIndex = group_id * g_NumOfStrandsPerThreadGroup + localStrandIndex;
+    localVertexIndex = (local_id - localStrandIndex) / g_NumOfStrandsPerThreadGroup;
+
+    strandType = GetStrandType(globalStrandIndex);
+    globalVertexIndex = globalStrandIndex * numVerticesInTheStrand + localVertexIndex;
+}
+
+void CalcIndicesInVertexLevelMaster(uint local_id, uint group_id, inout uint globalStrandIndex, inout uint localStrandIndex, inout uint globalVertexIndex, inout uint localVertexIndex, inout uint numVerticesInTheStrand, inout uint indexForSharedMem, inout uint strandType)
+{
+    indexForSharedMem = local_id;
+    numVerticesInTheStrand = (THREAD_GROUP_SIZE / g_NumOfStrandsPerThreadGroup);
+
+    localStrandIndex = local_id % g_NumOfStrandsPerThreadGroup;
+    globalStrandIndex = group_id * g_NumOfStrandsPerThreadGroup + localStrandIndex;
+    globalStrandIndex *= (g_NumFollowHairsPerGuideHair+1);
+    localVertexIndex = (local_id - localStrandIndex) / g_NumOfStrandsPerThreadGroup;
+
+    strandType = GetStrandType(globalStrandIndex);
+    globalVertexIndex = globalStrandIndex * numVerticesInTheStrand + localVertexIndex;
+}
+
+void CalcIndicesInStrandLevelTotal(uint local_id, uint group_id, inout uint globalStrandIndex, inout uint numVerticesInTheStrand, inout uint globalRootVertexIndex, inout uint strandType)
```

Comment:
I agree with you but would recommend keeping it as is this is the original index calc AMD code - this will help for future reference and changed.
Let me know if we can close this or you'd like me to change.

Reply:
Okay, if it's same as AMD then fine with keeping it.